### DISCUSSION TITLE
piecewise output same type as input

### DISCUSSION THREAD

**Discussion by hongwanliu** [jax-ml/jax#17178] [General] [answered]
Why does piecewise require the output to be the same as the input?

**Comment by jakevdp** [comment]
It's an implementation detail: the issue is that all functions in `piecewise` must return the same dtype, and casting to the input dtype was an easy way to make sure this condition was met.

Is this something that is blocking your use-case?

**Reply by hongwanliu** [reply]
Thanks. I can get around this for the case I was interested in, but it is a surprising requirement, and causes some inconvenience. I was trying to calculate the Riemann zeta function (which is very badly implemented in jax unfortunately: for negative, even integers it should be zero, but jax returns a value that is very much nonzero), and was hoping to pass integers and get floats as output.

**Reply by jakevdp** [reply, marked as answer]
Regarding the Zeta function, we just merged https://github.com/google/jax/pull/17144 which should make things much better.

That aside, I can understand why the current casting behavior of `jnp.piecewise` is not ideal. I'll look into changing that.

**Reply by hongwanliu** [reply]
Awesome. Thanks!

**Reply by jakevdp** [reply]
#17186 should fix the `piecewise` issue.

**Reply by jakevdp** [reply]
Actually, it looks like the casting behavior you mention matches that of numpy:
```python
In [1]: import numpy as np

In [2]: x = np.arange(10, dtype=int)

In [3]: np.piecewise(x, [x < 5, x > 5], [np.sin, np.cos]).dtype
Out[3]: dtype('int64')
```
Since the goal of `jax.numpy` functions is to provide the same API as `numpy`, but implemented in JAX, I think we should not merge #17186, but rather keep the behavior as-is. What do you think?

**Reply by hongwanliu** [reply]
Interesting. now that i’m aware of this for jax i’m ok either way, i just never needed piecewise in numpy whereas it’s much more important in jax.

**Reply by jakevdp** [reply]
For what it's worth, `piecewise` is just a thin numpy-style wrapper around [`lax.switch`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.switch.html) – if the numpy API convention is not working for you, I'd suggest using `lax.switch` directly.

Here's the core logic of the piecewise implementation: https://github.com/google/jax/blob/b4a628400f59b84a0a0bc47e828b389fc9c34bfd/jax/_src/numpy/lax_numpy.py#L5027-L5039

**Reply by jakevdp** [reply]
There is some discussion of this on NumPy's side in https://github.com/numpy/numpy/issues/24155; I'll keep an eye on that and update JAX if they change the API there.
